### ISSUE TITLE
Namespace quotas created by kubectl not being enforced

### ISSUE THREAD

**Issue by bentastic27** [rancher/rancher#25150]
Right now, project quotas are not enforced when namespaces are created outside of the Rancher UI/API. A customer is running into a situation where their users are creating namespaces with kubectl/kube api and the resource limits that should be imposed on these users are not being enforced.

They were attempting to workaround this by adding the `field.cattle.io/projectId: ${CLUSTER_ID}:${PROJECT_ID}` annotation, but since this is being done outside of Rancher. The project limits still do not work.

They would like for their users to be able to create namespaces and as long as they have the aforementioned annotation, the project limits will be applied to the namespace.

From the customer

> We are using kubectl to create namespaces in Rancher managed kubernetes clusters. These projects have project quotas and default namespace quotas set prior to namespace creation. When we need those namespaces to be assigned under a particular project, we set cattle annotations required to get the namespace assigned to a project.
> The namespaces are being created using defaults set in the project but, when we go on creating new namespaces, the project quotas are not really enforced and we are able to create unlimited namespaces with default quotas.

Also:

> there are no calculations happening so if the project and all its namespaces have already taken up all the resources… a new namespace can be provisioned using kubectl and be given the default resource constraint when it should actually be set to zero as a limit until someone updates the project limit.

gz#8077

**Comment by samjustus** [comment]
**Troubleshooting steps:**

Create a project, assign it a CPU/mem quota. After that you are still able to assign random ResourceQuotas to a namespace via Cluster Explorer/kubectl

**Comment by maxsokolovsky** [comment]
SURE 1681

**Comment by maxsokolovsky** [comment]
This is not a bug but a known limitation. Changing it requires a comprehensive design. Will review notes made by @cmurphy and some other information. Then we'll come up with an exact design and implementation.

**Comment by samjustus** [comment]
This is no longer relevant, as Rancher does enforce quota limits when the UI is bypassed. When users create namespaces in projects via kubectl, they need to specify an annotation to associate the namespace with a particular project (field.cattle.io/projectId: ${CLUSTER_ID}:${PROJECT_ID}).

Then Rancher creates the namespace and a ResourceQuota object associated with it, as it does when users create namespaces via UI.
Rancher does enforce namespace quota limits in the same way as when it receives a request from the UI, with one caveat [1].

I also tried creating a namespace that does not belong to any project. Initially, it does not have quota limits. Then I moved
it to a project with configured total limits and namespace default limits. The limits were properly applied to the new namespace.

All of the above applies to the upcoming v2.6.4, by the way. That being said, there might be edge cases where Rancher could improperly handle enforcement of resource quota limits. We might have not yet seen them, but we will fix them if they arise.

To conclude, Rancher does fully enforce resource quota limits when users make namespaces in projects via kubectl.

 

[1] When a new namespace is created via UI, Rancher grants the min(available, requested) amount of a resource.
For requests made with kubectl, a resource amount must be fully available in the project relative to what's already taken up, else zero is granted for that resource.
